Automatic recovery from permanent failures of etcd3 nodes

mumoshu opened the pull request:
closes #381, resolves #27

## Features

* Automatic recovery from any number of permanently failed etcd nodes
  * Currently, it is considered to be a "permanent failure" when health checks to an etcd member is failing longer than the threshold(10 seconds by default)

## Notable changes

* Upgraded etcd to v3
* kube-apiserver uses etcd3 API rather than the former etcd2 API for writing v3 keys
  * Although you can stick with v2 api + v3 etcd(then kube-apiserver writes v2 keys in etcd v3!), the current implementation of etcd snapshot/restore doesn't support etcd v2 data. That's why I made this change
  * etcd3 will be the default storage backend since k8s 1.6 anyways
* Wrote a lengthy `etcdadm` script to automate etcd member health checking, saving snapshots, recovering from up to `N/2` permanently failed nodes(=replace failed members one by one), recovering from more than `N/2` permanently failed nodes(=bootstrapping a brand-new cluster with the latest snapshot)

## Remaining issues

- [ ] TLS assets are currently located under `/etc/ssl/certs` rather than the former `/etc/etcd2/ssl`
  - This is due to my misunderstanding that `etcd-member.service` bundled in Container Linux for running etcd3 member doesn't support a convenient way to customize the location. It does expose `ETCD_SSL_DIR` to instruct an etcd3 rkt pod to mount `ETCD_SSL_DIR` to `/etc/ssl/certs` in the pod. So, by setting `ETCD_*_KEY_FILE` and `ETCD_*_CERT_FILE` prefixed with the `/etc/ssl/certs` regardless of where the TLS assets are in the host machine would just work.
- [ ] Etcd snapshots are currently saved in `<bucket from s3-uri>/<path from s3-uri>/kube-aws/clusters/<cluster name>/exported/stacks/etcd-snapshots/*`, which doesn't make sense
- [ ] Etcd data directories are still located at `/var/lib/etcd2` 😃 It should be `/var/lib/etcd` or `/var/lib/etcd3`
- [ ] `etcdadm` can't be modified as it is directly embedded from a kube-aws binary to `cloud-config-etcd`, without rendered at first by `kube-aws render` like other assets.

mumoshu commented:
Seeing bad certificate errors in the `etcdadm-check.service` on etcd nodes.
Probably I confused etcd peer key/cert with client key/cert :)
Fixing...

mumoshu commented:
Try triggering a permanent failure by running the below in one of your etcd nodes:

```
sudo bash -c 'find /var/lib/etcd2/member/ -exec bash -c "echo corrupting \$1; echo poison > \$1" -- {} \;; systemctl restart etcd-member'
```

And see how it works by tailing logs:

```
journalctl -f -u etcdadm-reconfigure
```

If you've done the above step to all the nodes at once, a disaster recovery process would be triggered i.e. all the members are recovered from the latest snapshot (if any. snapshots are taken since 120sec after each node started)

If you'd like to replace a permanently failed etcd node at any time you like, run:

```
sudo bash -c 'systemctl stop etcd-member; set -a; source /etc/etcd-environment; source /var/run/coreos/etcadm-environment; set+a; /opt/bin/etcdadm replace; systemctl start etcd-member'
```

mumoshu commented:
Fixed various bugs

redbaron commented:
>  Etcd snapshots are currently saved in <bucket from s3-uri>/<path from s3-uri>/kube-aws/clusters/<cluster name>/exported/stacks/etcd-snapshots/*, which doesn't make sense

I kinda like that there is no additional parameter to setup :)

redbaron commented on etcdadm/etcdadm:
```diff
@@ -0,0 +1,1559 @@
+  date +%s
+}
+
+_run_as_root() {
```
This might work too:
```bash
_run_as_root=

if [[ "$EUID" > 0 ]]; then
  _run_as_root=sudo
fi
```

mumoshu replied:
Elegant 🌹

redbaron commented on etcdadm/etcdadm:
```diff
@@ -0,0 +1,1559 @@
+  exit 1
+}
+
+_array_join() {
```
```
printf '%q ' "${@}" 
```

mumoshu replied:
Perfect!
Your impl seems to work even for a complex case like this 👍 
```
./test.sh sudo bash -c "bash -c 'echo \"foobar baz\"'" | bash
#=> foobar baz
```

redbaron commented on etcdadm/etcdadm:
```diff
@@ -0,0 +1,1559 @@
+  echo "docker://$(config_awscli_docker_image)"
+}
+
+config_aws_region() {
```
why functions which print values? can't values be just set?

mumoshu replied:
I guess these functions especially had been useful only when I had not set `set -e` :)
I'm ok with making them values as long as it can emit an undefined-variable-error where necessary.
I'll try later 👍

redbaron commented on etcdadm/etcdadm:
```diff
@@ -0,0 +1,1559 @@
+  echo "https://github.com/coreos/etcd/releases/download/$v/etcd-$v-linux-amd64.aci"
+}
+
+cluster_member_indices() {
```
every time it is used `seq $(config_member_count)` can be used instead

mumoshu replied:
Thanks!
As it is starting from 0, would it be something like `seq 0 (($(config_member_count) -1`, or is there anything better?

redbaron replied:
you are right, `seq 0 $(( $(config_member_count) -1 ))` is good enough probably :)

alternative is `for x in {0..$(( $(config_member_count) -1 ))}`

mumoshu replied:
Thanks.
I tried the alternative but it seems to not work(?)

```
core@core-01 ~ $ num=3; for x in {0..$(( num -1 ))}; do echo $x; done
{0..2}
core@core-01 ~ $ num=3; for x in {0..3}; do echo $x; done
0
1
2
3
```

redbaron replied:
well, it is not really an alternative then :))

redbaron commented on etcdadm/README.md:
```diff
@@ -0,0 +1,104 @@
+* rkt
+* docker
+* systemd
+* bash 3+
```
`jq`

mumoshu replied:
Good catch! Thanks

mumoshu replied:
Added.

redbaron commented on etcdadm/etcdadm:
```diff
@@ -0,0 +1,1559 @@
+    describe_instances=$(_awscli_command ec2 describe-instances --filter Name=tag:kube-aws:role,Values=etcd Name=tag:KubernetesCluster,Values=$KUBERNETES_CLUSTER)
+    $describe_instances | jq -r '[ .Reservations[].Instances[] ] | length'
+  else
+    local f
```
if file cannot be empty, the can be written shorter
```
else
  cat $(tester_num_running_nodes_file) 2>/dev/null || echo 0
fi
```

redbaron commented on etcdadm/etcdadm:
```diff
@@ -0,0 +1,1559 @@
+  return 1
+}
+
+cluster_majority() {
```
can be written shorter:
```
cluster_majority() {
  echo $(( $(config_member_count) / 2 + 1 ))
}
```

but again, given that it is a static value, it can't it be just set and used elsewhere instead of specifying and  calling a function?

mumoshu replied:
It was written as so to prevent successful `echo` even when `config_member_count` is failed.
However, making it just a variable would be better as I have `set -o nounset` now 👍

mumoshu commented:
> I kinda like that there is no additional parameter to setup :)

It will be `<bucket from s3-uri>/<path from s3-uri>/kube-aws/clusters/<cluster name>/etcd-snapshots/*` 😄

redbaron commented on core/controlplane/config/templates/cloud-config-etcd:
```diff
@@ -24,15 +24,120 @@ coreos:
+        [Service]
+        Type=oneshot
+        RemainAfterExit=yes
+        RestartSec=5
```
does systemd restart `oneshot` services now?

mumoshu replied:
No! Good catch.
Btw, I wanted the reconfiguration to retry until it eventually succeeds. However, setting it to  `Type=simple` breaks `etcd-member.service`. I hope there's any way to easily make it restart-able.

redbaron commented on core/controlplane/config/templates/cloud-config-etcd:
```diff
@@ -24,15 +24,120 @@ coreos:
           content: |
             {{- if .TLSAssetsEncryptionEnabled}}
             [Unit]
             Wants=decrypt-tls-assets.service cfn-etcd-environment.service
```
don't we need `cfn-etcd-environment.service` all the time?

mumoshu replied:
It is required to dynamically populate `/etc/etcd-environment` which is then read by etcd?

redbaron replied:
that is what I thought too,but it is inside `if` statement , shouldn't it be outside?

`[Unit]` sees to be inside `if` too, which doesn't look right

mumoshu replied:
Certainly! I didn't get that until now.
Perhaps it was luckily working when it had changed in #390, by the existence of `Before=format-etcd2-volume.service` of `cfn-etcd-environment.service`. Anyway, it should be fixed.

redbaron commented on core/controlplane/config/templates/stack-template.json:
```diff
@@ -331,6 +331,33 @@
+                  "Resource": "arn:{{.Region.Partition}}:s3:::{{$.EtcdSnapshotsS3Bucket}}",
+                  "Condition": {
+                    "StringLike": {
+                      "s3:prefix": "{{$.EtcdSnapshotsS3Prefix}}/*"
```
nice attention to details!

redbaron commented on core/controlplane/config/templates/stack-template.json:
```diff
@@ -461,6 +488,11 @@
             "Value": "{{$.ClusterName}}-{{$.StackName}}-kube-aws-etcd-{{$etcdIndex}}"
+          },
+          {
+            "Key": "kube-aws:role",
```
can't it use Name tag above? it kind find its own tag and then strip index to find other etcd instances

mumoshu replied:
That's possible. I just prefer the dedicated key though because then it won't be affected by changes to `Name` and the etcdadm script becomes a bit simpler

redbaron replied:
reason I asked is because there can be 10 Tags on a resource and company policies might require 4-5 of them to be set, so no much room for new tags :) I'd leave it simple then, until it becomes a problem

mumoshu replied:
argh 😄  certainly it should be considered in (near) future..

mumoshu commented:
A few E2E test cases are failing, probably due to etcd3 storage provider?

```
Summarizing 4 Failures:

[Fail] [k8s.io] Pods Delete Grace Period [It] should be submitted and removed [Conformance]
/go/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/pods.go:176

[Fail] [k8s.io] Kubectl client [k8s.io] Kubectl run --rm job [It] should create a job from an image, then delete the job [Conformance]
/go/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/kubectl.go:1278

[Fail] [k8s.io] Pods [It] should be submitted and removed [Conformance]
/go/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/common/pods.go:250

[Fail] [k8s.io] SchedulerPredicates [Serial] [It] validates resource limits of pods that are allowed to run [Conformance]
/go/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/scheduler_predicates.go:200

Ran 117 of 474 Specs in 4953.109 seconds
FAIL! -- 113 Passed | 4 Failed | 0 Pending | 357 Skipped --- FAIL: TestE2E (4953.12s)
FAIL

Ginkgo ran 1 suite in 1h22m33.44099363s
```

redbaron commented:
it would mean that not everything in kube works with etcd3 :) maybe tests are just flaky

mumoshu commented:
I hope so 😄 Anyways, these cases seems to be consistently failing so far

mumoshu commented:
If E2E doesn't pass with the combination of k8s 1.5.4 and etcd3, should we postpone this until k8s v1.6 is released?
Since k8s 1.6, the default storage provider etcd3 and therefore the flakiness should be fixed after the release.

redbaron commented on etcdadm/etcdadm:
```diff
@@ -6,72 +6,70 @@ unset ETCDCTL_ENDPOINT
 
+set -o nounset
+set -o errexit
+set -o pipefail
```
this is dangerous, it makes commands like `cat BIGFILE |  grep -q .` to fail, as `cat` receives `SIGPIPE` when `grep` exits and therefore exits with non 0 code, causing whole expression to fail

mumoshu replied:
Good to know! Thanks.
Are there any better ones than the current combination of options?

mumoshu commented on etcdadm/etcdadm:
```diff
@@ -0,0 +1,1621 @@
+    elif cluster_is_failing_longer_than_limit; then
+      _info "all the nodes for etcd members are running but cluster has been unhealthy for a while, which means cluster is now in disaster recovery process. searching for a etcd snapshot to recover this member"
+      member_bootstrap
+    else
```
Perhaps this block results in the last etcd node to miss restoring itself from an etcd snapshot, if any?

We can probably change this if-else block to just:

```bash
f (( running_num < total_num )); then
  _info "only ${running_num} of ${total_num} nodes for etcd members are running, which means cluster is still in bootstrap process. searching for a etcd snapshot to recover this member"
elif (( running_num == total_num )); then
  _info "all the ${running_num} of ${total_num} nodes for etcd members are running. Almost certainly this node is the last one?"
elif cluster_is_failing_longer_than_limit; then
  _info "all the nodes for etcd members are running but cluster has been unhealthy for a while, which means cluster is now in disaster recovery process. searching for a etcd snapshot to recover this member"
else
  _panic "[bug] invalid state!"
fi
member_bootstrap
```

whereisaaron commented:
1.6 is out! etcdv3 for all :)

mumoshu commented:
Yay! 😆
@whereisaaron Btw, are you ok with dropping etcd v2 support form kube-aws after this?
That is a major headache for me as of now 😢

mumoshu commented:
TODO: A configuration key in cluster.yaml to enable/disable the automatic disaster recovery feature?

whereisaaron commented:
@mumoshu I don't have any need to support etcd v2 and v3 in the same version of `kube-aws`. Since `kube-aws` versions are often not backwards compatible, I use only a specific version of `kube-aws` to create and update each cluster.

I suggest you tie k8s v1.6 and etd v3 together, and that version of `kube-aws` support only k8s v1.6+ with etcd v3. And stick with etc v2 for k8s 1.5 and earlier.

The automated disaster recovery sounds great to me, but I imagine it might not be everyone's taste? I suggest there are three levels of enabled:
1. Do nothing
2. Snapshots to S3
3. Snapshots to S3 with auto-recovery from snapshot

mumoshu commented on etcdadm/README.md:
```diff
@@ -0,0 +1,109 @@
+
+## Other notes
+
+Although it might be a bad idea, `etcadm` is implemented in bash (for now) due to the fact it requires access to rkt, docker, systemd and etcd data dir on a host machine to reconfigure an etcd member.
```
typo: "etcadm" should be "etcdadm"

mumoshu commented:
@whereisaaron Thanks for the suggestion 👍 

 I've added `etcd.snapshot.automated: true` for taking snapshots periodically, and  `etcd.disasterRecovery.automated: true` for enabling the automated disaster recovery feature.

mumoshu commented:
I've also added `etcd.version` for choosing either `2` or `3` as a temporary step until we settle to k8s 1.6 and eventually drop the support for etcd2.

mumoshu commented:
Due to my laziness, I kept the change from `/etc/etcd/ssl/` to `/etc/ssl/certs` as the directory containing enc files and the decrypted files as-is for now. Making it other than `/etc/ssl/certs` makes paths among the host and the etcd3 container inconsistent and complicate generation of several env files.

mumoshu commented:
Just for the record, conformance test is still failing with k8s 1.5.5 + etcd3 as is well known.

```
Summarizing 5 Failures:

[Fail] [k8s.io] SchedulerPredicates [Serial] [BeforeEach] validates that NodeSelector is respected if matching [Conformance]
/go/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/scheduler_predicates.go:78

[Fail] [k8s.io] Service endpoints latency [It] should not be very high [Conformance]
/go/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/service_latency.go:114

[Fail] [k8s.io] SchedulerPredicates [Serial] [BeforeEach] validates resource limits of pods that are allowed to run [Conformance]
/go/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/scheduler_predicates.go:78

[Fail] [k8s.io] Pods [It] should be submitted and removed [Conformance]
/go/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/common/pods.go:250

[Fail] [k8s.io] Pods Delete Grace Period [It] should be submitted and removed [Conformance]
/go/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/pods.go:176

Ran 117 of 474 Specs in 4897.388 seconds
FAIL! -- 112 Passed | 5 Failed | 0 Pending | 357 Skipped --- FAIL: TestE2E (4897.42s)
FAIL
```

mumoshu commented:
As k8s 1.5.5 + etcd2 is still passing conformance test on this branch, I'll merge this shortly and start testing with k8s 1.6 while enabling the optional etcd3 support.

mumoshu commented:
The integration test suite for etcdadm passed:

```
$ cd etcdadm
$ make test-integration
*snip*
finished testing recovery from permanent member failures

all the tests passed. tailing log...
```

mumoshu commented:
I'm not satisfied with the overall quality of this work but it seems to be working anyway.
Let's go forward.
